tracer/test/Datadog.Trace.Tests/Sampling/SpanSamplerTests.cs

```diff
@@ -0,0 +1,210 @@
+            var rules = new List<SpanSamplingRule>() { rule1 };
+            var sampler = new SpanSampler(rules);
+            var span = new Span(new SpanContext(5, 6, null, serviceName: "service-name"), DateTimeOffset.Now) { OperationName = "operation-name" };
+
+            sampler.MakeSamplingDecision(span);
+
+            Assert.Equal(expectedRuleRate.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.RuleRate));
+            Assert.Null(span.Tags.GetTag(Tags.SingleSpanSampling.MaxPerSecond));
+            Assert.Equal(expectedSamplingMechanism.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism));
+        }
+
+        [Fact]
+        public void Tags_ShouldBe_Correct()
+        {
+            var expectedRuleRate = 0.99f;
+            var expectedMaxPerSecond = 500.0f;
+            var expectedSamplingMechanism = 8; // SamplingMechanism.SpanSamplingRule
+            var rule1 = new SpanSamplingRule("service-name", "operation-name", 0.99f, 500.0f);
+            var rules = new List<SpanSamplingRule>() { rule1 };
+            var sampler = new SpanSampler(rules);
+            var span = new Span(new SpanContext(5, 6, null, serviceName: "service-name"), DateTimeOffset.Now) { OperationName = "operation-name" };
+
+            sampler.MakeSamplingDecision(span);
+
+            Assert.Equal(expectedRuleRate.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.RuleRate));
+            Assert.Equal(expectedMaxPerSecond.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.MaxPerSecond));
+            Assert.Equal(expectedSamplingMechanism.ToString(), span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism));
+        }
+
+        [Fact]
+        public void Allow_All()
+        {
+            var allowAllRule = new SpanSamplingRule("*", "*");
+            var rules = new List<SpanSamplingRule>() { allowAllRule };
+            var sampler = new SpanSampler(rules);
+
+            RunSamplerTest(sampler, 500, expectedAutoKeepRate: 0, expectedUserKeepRate: 1, acceptableVariancePercent: 0);
+        }
+
+        [Fact]
+        public void Allow_None_SamplingRate()
+        {
+            var allowNoneRule = new SpanSamplingRule("*", "*", 0.0f);
+            var rules = new List<SpanSamplingRule>() { allowNoneRule };
+            var sampler = new SpanSampler(rules);
+
+            RunSamplerTest(sampler, 500, expectedAutoKeepRate: 1, expectedUserKeepRate: 0, acceptableVariancePercent: 0);
+        }
+
+        [Fact]
+        public void Allow_Half_SamplingRate()
+        {
+            var allowHalfRule = new SpanSamplingRule("*", "*", 0.5f);
+            var rules = new List<SpanSamplingRule>() { allowHalfRule };
+            var sampler = new SpanSampler(rules);
+
+            RunSamplerTest(sampler, 500, expectedAutoKeepRate: 0.5f, expectedUserKeepRate: 0.5f, acceptableVariancePercent: 0.1f);
+        }
+
+        // TODO rate limiter test? didn't seem to be reliable though
```

Comment:
If the aim is to really UnitTest the `SpanSampler` code, you could imagine mocking the SpanRateLimiter and using an internal .ctor to set it. That should be relatively easy and would remove the unreliable part :)

Reply:
AFAICT, the `SpanSampler` code itself doesn't directly depend on `SpanRateLimiter`, so doesn't really seem necessary to have these tests to me. The `SpanSampler` code itself is pretty simple